### PR TITLE
Try/publish in site editor

### LINKED ISSUE #49873: Site Editor: enable option to publish pages
When creating new pages from a navigation block, you can create draft pages. However, within the site editor, there's no way to publish pages, despite being able to edit them directly. This relates to the prior issue reported here around confusion with the draft state: https://github.com/WordPress/gutenberg/issues/49871 This was found as I was working on the next call for testing for the FSE Outreach Program.

### PR DESCRIPTION
**Pull request by SaxonF**
## What?
Adds ability to set status, publish date and password in site editor.

image

image

## Why?
Resolves https://github.com/WordPress/gutenberg/issues/49873

## How?
WIP. 

## Testing Instructions
1. Open site editor
2. Go to `pages` and select a page to edit
3. Edit the page and open up the page inspector and notice the new summary panel
4. Try changing the status from draft to published
5. Try updating the publish date to a point in the future and notice status changes to scheduled
6. Try updating the publish date to a point in the past and notice status changes back to published
7. Hit save and check to see if status change is persisted

### CODE REVIEW

**Comment by jameskoster** [comment]
It seems quite likely that one would want to edit the status from both the Page Details panel, and the Page Inspector in the Editor. In either case, status, visibility, and password protection don't seem like details you will edit with all that much frequency.

These assumptions make me wonder whether a modal could work well for this edit UI. The same result is produced, regardless of entry point:

Screenshot 2023-06-12 at 14 56 30

It could even work as a drilldown, if we wanted to go in that direction.

What do you think?

**Comment by SaxonF** [comment]
@jameskoster I updated PR to use modal initially just triggered via page inspector. One bit of jank is that it doesn't "save" the status change until a user hits save . We could save but would need to prompt to save other details before. Have a play and see what you think.

**Comment by jameskoster** [comment]
Nice.

I don't actually mind the jank, at least it's consistent with the post editor / other meta changes.

We might need to include the calendar for all statuses except Draft and Pending, otherwise you cannot update the publish date.

File: packages/edit-site/src/components/sidebar-edit-mode/page-panels/page-status.js [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+							<BaseControl
+								hideLabelFromVision
+								label={ __( 'Status' ) }
+								id={ `edit-site-change-status__status` }
+								className={ 'components-radio-control' }
+							>
+								<VStack spacing={ 3 }>
+									{ STATUS_OPTIONS.map( ( option, index ) => (
+										<div
+											key={ `edit-site-change-status__status-${ index }` }
```
Comment (kevin940726) [inline packages/edit-site/src/components/sidebar-edit-mode/page-panels/page-status.js]: Nit: It's generally a bad practice to use indices as keys. We can use `value` here to better convey meaning.


```suggestion
											key={ option.value }
```

File: packages/edit-site/src/components/sidebar-edit-mode/page-panels/page-status.js [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+											className="components-radio-control__option with-hint"
+										>
+											<input
+												id={ `edit-site-change-status__status-${ index }` }
```
Comment (kevin940726) [inline packages/edit-site/src/components/sidebar-edit-mode/page-panels/page-status.js]: Non-blocking: We can use [`useId`](https://react.dev/reference/react/useId) here to generate a unique id and append `value` afterward. Something like:

```js
const id = useId();

<input id={ `${id}-${option.value}` } />
```

File: packages/edit-site/src/components/sidebar-edit-mode/page-panels/page-status.js [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+												className="components-radio-control__input"
```
Comment (kevin940726) [inline packages/edit-site/src/components/sidebar-edit-mode/page-panels/page-status.js]: Why don't we use the `<RadioControl>` component directly? I'm not sure but I think it could be a bad practice to use their class names outside of the components package. 🤔 c.c. @mirka @ciampo

File: packages/edit-site/src/components/sidebar-edit-mode/page-panels/page-status.js [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+								/>
+								{ showPassword && (
+									<TextControl
+										onChange={ ( value ) =>
+											saveStatus( { password: value } )
+										}
+										value={ password }
+										/* eslint-disable jsx-a11y/no-autofocus */
+										autoFocus
+										/* eslint-enable jsx-a11y/no-autofocus */
```
Comment (kevin940726) [inline packages/edit-site/src/components/sidebar-edit-mode/page-panels/page-status.js]: This could be an accessibility issue. Might want to get a second eye from an accessibility expert.

File: packages/edit-site/src/components/sidebar-edit-mode/page-panels/page-status.js [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+										placeholder={ __(
+											'Enter a secure password'
+										) }
+										type="password"
```
Comment (kevin940726) [inline packages/edit-site/src/components/sidebar-edit-mode/page-panels/page-status.js]: This throws a warning in Chrome:

> [DOM] Password field is not contained in a form: (More info: (https://goo.gl/9p2vKq)

There is some great info in the link that we can take into consideration. Not sure how much is relevant here though!

File: packages/edit-site/src/components/sidebar-navigation-screen-page/style.scss [hunk 6]
```diff
@@ -77,4 +77,3 @@
 		color: $gray-200;
 	}
 }
-
```
Comment (kevin940726) [inline packages/edit-site/src/components/sidebar-navigation-screen-page/style.scss]: Possibly unnecessary change due to editor setup? 😆

File: packages/edit-site/src/components/sidebar-edit-mode/page-panels/page-status.js (hunk 4 above)
Comment (kevin940726) [reply packages/edit-site/src/components/sidebar-edit-mode/page-panels/page-status.js]: This also means that if you've set a password and closed the popover, when you reopen it again, the focus will be moved to here instead of the first focusable element. This is probably not what we want?

File: packages/edit-site/src/components/sidebar-edit-mode/page-panels/page-status.js [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+	const { createErrorNotice } = useDispatch( noticesStore );
+
+	const [ popoverAnchor, setPopoverAnchor ] = useState( null );
+	// Memoize popoverProps to avoid returning a new object every time.
+	const popoverProps = useMemo(
+		() => ( {
+			// Anchor the popover to the middle of the entire row so that it doesn't
+			// move around when the label changes.
+			anchor: popoverAnchor,
+			placement: 'bottom-end',
```
Comment (kevin940726) [inline packages/edit-site/src/components/sidebar-edit-mode/page-panels/page-status.js:100]: The popover doesn't have a descriptive name for screen readers, users might not even know that they land in a popover. Maybe we should add `aria-label` or `aria-labelledby` here.

The other thing is that `Popover` traps the focus but it doesn't set the `dialog` role. This could be a bigger problem in general of the component design and I personally think it's an accessibility bug. We could set a `role: 'dialog'` here but I'm not sure if it's the right move. Possibly better to consult an accessibility expert on this.

It's still usable thought just a bit confusing so I don't think this should be blocking 👍

**Review by kevin940726 — COMMENTED** [review]
Thanks! Left some suggestions but most of them are not blocking and can be improved in future PRs 👍 . Just jotting them down here to keep track of them in case I forget 😝 .

**Comment by SaxonF** [comment]
@kevin940726 Thanks for the review! I'll fix up your suggestions first thing tomorrow and hopefully we can get this merged later in the day :)

File: packages/edit-site/src/components/sidebar-edit-mode/page-panels/page-status.js (hunk 3 above)
Comment (SaxonF) [reply packages/edit-site/src/components/sidebar-edit-mode/page-panels/page-status.js]: @kevin940726 Ideally we would but in this case we have an added help/hint text for each _option_ that the current RadioControl doesn't support (as far as I'm aware). We use the same approach for the existing PostVisibility component. 

Perhaps we could open up a separate PR within the components package that adds help support to options?

File: packages/edit-site/src/components/sidebar-edit-mode/page-panels/page-status.js (hunk 4 above)
Comment (SaxonF) [reply packages/edit-site/src/components/sidebar-edit-mode/page-panels/page-status.js]: Good point. I've updated to only focus input when password is empty.

File: packages/edit-site/src/components/sidebar-edit-mode/page-panels/page-status.js (hunk 5 above)
Comment (SaxonF) [reply packages/edit-site/src/components/sidebar-edit-mode/page-panels/page-status.js]: @kevin940726 would you recommend wrapping the radio inputs and password field in a form? It's tricky because we aren't actually "submitting" anything until site editor changes are saved via the save button. I'm not sure what the best practice here is.

File: packages/edit-site/src/components/sidebar-edit-mode/page-panels/page-status.js (hunk 7 above)
Comment (SaxonF) [reply packages/edit-site/src/components/sidebar-edit-mode/page-panels/page-status.js:100]: Added an `aria-label` for the time being since I don't want to modify `InspectorPopoverHeader` or `PublishDateTimePicker` as part of this PR to add ID to title. 

> The other thing is that Popover traps the focus but it doesn't set the dialog role. This could be a bigger problem in general of the component design and I personally think it's an accessibility bug. We could set a role: 'dialog' here but I'm not sure if it's the right move. Possibly better to consult an accessibility expert on this.

Let's create a separate issue referencing the Dropdown and Popover components in wordpress/components to review any accessibility concerns. [Modal](https://wordpress.github.io/gutenberg/?path=/story/components-modal--default) has a `title` prop that uses `aria-labelledby`.

File: packages/edit-site/src/components/sidebar-edit-mode/page-panels/page-status.js (hunk 3 above)
Comment (kevin940726) [reply packages/edit-site/src/components/sidebar-edit-mode/page-panels/page-status.js]: I see! Could we just render the hint inside `label`?

```js
<RadioControl
  {...props}
  label={
    <>
      <div>{option.label}</div>
      {option.hint && <Text variant="muted">{option.hint}</Text>
    </>
  }
/>
```

This has a nice side effect of allowing clicking on the hint the select the radio button too. We would still need to update the type of the original component though.

File: packages/edit-site/src/components/sidebar-edit-mode/page-panels/page-status.js (hunk 5 above)
Comment (kevin940726) [reply packages/edit-site/src/components/sidebar-edit-mode/page-panels/page-status.js]: I'm not sure either, TBH! I guess it's still fine to wrap the whole popover in a dummy `<form>` just to please chrome?

File: packages/edit-site/src/components/sidebar-edit-mode/page-panels/page-status.js (hunk 3 above)
Comment (SaxonF) [reply packages/edit-site/src/components/sidebar-edit-mode/page-panels/page-status.js]: That feels a bit cleaner. I still had to add some custom CSS to account for spacing and hint text alignment.

**Comment by SaxonF** [comment]
@kevin940726 ready for a re-review please!

File: packages/edit-site/src/components/sidebar-edit-mode/page-panels/page-status.js [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+	const { createErrorNotice } = useDispatch( noticesStore );
+
+	const [ popoverAnchor, setPopoverAnchor ] = useState( null );
+	// Memoize popoverProps to avoid returning a new object every time.
+	const popoverProps = useMemo(
+		() => ( {
+			// Anchor the popover to the middle of the entire row so that it doesn't
+			// move around when the label changes.
+			anchor: popoverAnchor,
+			ariaLabel: __( 'Change status' ),
```
Comment (kevin940726) [inline packages/edit-site/src/components/sidebar-edit-mode/page-panels/page-status.js]: I think it should be `aria-label`? The same goes for the publish-date popover :).

**Review by kevin940726 — APPROVED** [review]
Works for me locally. Just a small suggestion left! None of the issues mentioned above should be blocking so I'm approving this 👍
